### ISSUE TITLE
Left-aligned labels overlap fields on narrow screens

### ISSUE THREAD

**Issue by mattots** [torchbox/wagtail#1442]
This is particularly a problem inside blocks where the fieldset width is reduced.

![image](https://cloud.githubusercontent.com/assets/301729/8358822/63667ea6-1b5a-11e5-8f33-1a5923771193.png)

(I've added some additional CSS to show the dashed dividers between blocks and make the '+' buttons always visible, but other than that the styles are out-of-the-box Wagtail v1.0rc1.)

**Comment by davecranwell** [comment]
To an extent this can't be fixed. If we widen the space for the label in that "items" list it solves this specific instance of the problem, but if you later add a *second* level of nesting e.g another structblock inside each  "items" list item,  we're in the same position again. At some point we have to say "we only support nesting [x] levels deep".

That said, two levels of nesting won't be out of the ordinary so we should try and make this work better, although it will make those input boxes much narrower.

**Comment by mattots** [comment]
I wonder whether this is a good reason for having the labels above the fields rather than to the left?

**Comment by davecranwell** [comment]
At one point long ago labels were above inputs, but this made for very long editing pages. It would improve the situation here, but it wouldn't solve it completely. 

Our issue is essentially that we've created the possibility of an infinitely deep hierarchy, displayed using horizontal indenting inside a finite-width space.

**Comment by mattots** [comment]
Perhaps the problem here isn't so much the labels being left-aligned (although that does certainly create potential for problems) and more to do with the fact that the nested items reduce in width, to visually 'fit inside' the parent field. I can see the reason for doing this, but I wonder if there's not another way of indicating the hierarchy that avoids the problems identified - for example, placing borders around the nested items somehow?

**Comment by davecranwell** [comment]
The indent on the left is fairly obvious but the equal indent on the right is to allow for `help_text` to be rendered in the same relative position as it is for the page `title` field, without overlapping the help text of the parent block.

I'm thinking we need to move the help text, but this opens a can of worms. Currently the help text on the right pleasantly fades in when you hover over. It gets out of your way the rest of the time and its length is immaterial. However if help text is placed inline with the fields - e.g above each block - then that fading poses a problem. Faded-out text help text would result in a bizarre gap (of any possible height, depending how much help you write) between objects, and the editing interfaces starts to become an exercise in scrolling. If you combat this by making the help text permanently visible, the clean aesthetic suffers. 

Inline help text also creates consistency/aesthetic issues when used on fields that use the "full" class to display without input borders, like the `title` field.

Perhaps if the help text were in a small modal, absolutely positioned to overlap inputs, but had to be manually toggled to appear? That still doesn't solve the infinite nesting problem, but one thing at a time!

**Comment by mattots** [comment]
Although aesthetically speaking I like the idea of the help text only fading in when you hover over the field, from a usability point of view I wonder if this is perhaps not the best approach.

I don't think we can assume that the a user will have necessarily clicked on or hovered over a field before needing to see the help text. They may prefer to scan the page and make decisions based on the information provided, especially if the help text is actually telling the user more about what a particular field is for.

I think that having the help text always visible, above or below the field, is preferable to the current arrangement.

**Comment by davecranwell** [comment]
Certainly for the amount that `help_text` gets used on blocks (very little),  having help text permanently visible above each block might work better for most of the FieldPanels in Wagtail.

Visible help text would still be an aesthetic challenge with the design of StreamField though. You've chosen to manually alter the styles to permanently show the "+" buttons and mark the boundaries of each block, which is fine, but it's not our vision for streamfield.

We saw it more akin to how writing a Google Docs works: a magically blank space in which you can write whatever you choose seemingly without borders or differentiation of types of field.

We use it for creating rich, long, free-form content of the sort you see in major newspaper websites - we rarely use StructBlock or indeed any of the blocks where field borders/labels are visible. And in this situation, where almost none of the fields have help text or borders, it does already closely resemble the Google Docs ideal we had in mind. If we added permanently visible help text it would detract somewhat.

Ultimately we're going to have to balance the contrived needs of our own clients, with the arbitrary needs of everyone else who uses Wagtail.

**Comment by alexgleason** [comment]
I think the help text in Wagtail should be on the bottom instead of on the side.

Check out how help text is handled in Craft's matrix (similar to StreamField). It's very small and appears under the field. I think this is better because help text isn't always given, and when it is it takes up very little space. Then you could nest fairly deep without the fields shrinking so much.

![screenshot from 2015-09-23 09 57 27](https://cloud.githubusercontent.com/assets/3639540/10048638/88098706-61d9-11e5-954d-c2cf3e3c842b.png)

This won't solve the labels, but it will solve [my issue](https://github.com/torchbox/wagtail/issues/1716), which is that the editor doesn't have enough space in the field to write long content due to the empty help text sections on the right.

**Long term and theoretical thoughts:**

Note that the labels are also above the fields in Craft, which you pointed out would make the page very long, but I think that could be accounted for by shrinking the size of the fields. And in general I think people are willing to scroll. Websites can go up and down infinitely, but the width at any time is undefined so in my view it's better to embrace this by stacking things. Then your mobile view and desktop view are more synonymous, which also makes maintenance better because there are less media queries, so less code to deal with. It's also worth noting that Wagtail's editor admin already does have labels above fields for regular Django fields, so there's already an established pattern for it.

![screenshot from 2015-09-23 10 18 49](https://cloud.githubusercontent.com/assets/3639540/10049292/7f8835f2-61dc-11e5-8550-b4ebb6c2c70f.png)

Basically what I'm saying is, I think the benefits of stacking the labels outweighs making people scroll more, especially since the scrolling could be minimized by shrinking the size of the fields and in extreme cases the content can be separated into tabs. As you said, someone nesting StreamField too deep is probably using it in a way that wasn't intended, but similarly someone adding an insane amount of content to one panel should probably consider breaking it up better.

**Comment by tomdyson** [comment]
Thanks for this thoughtful feedback, @alexgleason. It's definitely something for us to consider.

**Comment by davecranwell** [comment]
It should be noted that help text for _fields_ does appear beneath the field. It's help text describing the _group_ of fields e.g for `MultiFieldPanel`s which appears on the right. This kind of help text is to describe a wider purpose, perhaps a note to the editor about how several fields combined to achieve a result. Its a fairly unique level of editorial help as CMSs go, but its something even we use rarely so worth reviewing.

**Comment by SalahAdDin** [comment]
@alexgleason Which template or style have the project in your images?

**Comment by alexgleason** [comment]
They are screenshots of Craft CMS. buildwithcraft.com/

**Comment by SalahAdDin** [comment]
Oh, seems elegant, but, only wagtail bitches, jajajajajjaja

**Comment by Emkaytoo** [comment]
**TL;DR:** if you have ever used a multitrack editor for sound or music editing, you're aware of the **solo** button. This is a solo button for Wagtail.

---
I gave some thought to this after @alexgleason brought it to my attention when he was building out a site for us in Wagtail. I see this issue as a conflict between two problems: a content hierarchy that's too vertical/deep and a content management experience wanting to be simplified. There are a two immediate solutions I can think of. One is technical and is described below, and the other is human and is for another conversation.

**Focus state:**
One solution is a focus feature. It's impossible to know every conceivable content hierarchy that clients and consultants will think of prior to developing a product. One way around that is a *focus* state that allows the content manager to rid themselves of the hierarchy altogether to focus on a particular block. For example:

![wagtail-current](https://cloud.githubusercontent.com/assets/6294225/11483425/e7146c06-976d-11e5-9479-e8afcf46f0af.jpg)

The above depicts the default display state of the system. As @davecranwell implied, a relatively flat on-page content schema would be fine given that the user is on a desktop (which I would guess is more likely), or at least a tablet when managing content. But what of multiple levels like [this blog post](http://pennebaker.com/journal/not-another-change-initiative/), which has nested `<ol>` and `<ul>` elements.

My thoughts are to add a sidebar with an indication of which sub-blocks are present in the immediate block, as shown below:

![wagtail-rest](https://cloud.githubusercontent.com/assets/6294225/11483398/c24f1b14-976d-11e5-8ba9-551293e02935.jpg)  

The headline and list blocks are noted in the sidebar. This is a simple model, but I think it works for more complex ones as well. If there was a sub-block inside each list item, it would list each sub-block as an item in the sidebar. Conceptually, it's kind of like showing the DOM for each block and sub-block type. Not each field, but each block. 

When a user clicks on a block in the side bar, as shown below, that block is given a *focus* state. 

![wagtail-focus](https://cloud.githubusercontent.com/assets/6294225/11483374/a89b4f8a-976d-11e5-9388-7ddea239070a.jpg)

I imagine this being some javascript that adds a `focus` class for the clicked item, giving that sub-block a set of full-width display properties. This would allow the content manager to focus on that item, and click off of it when s/he is done or needs to see it in hierarchy.

**Put another way:** if you have ever used a multitrack editor for sound or music editing, you're aware of the **solo** button. This is a solo button for Wagtail.

**Comment by davecranwell** [comment]
I like the solo button idea. For a while now I've been mulling an idea of having a navigation bar somewhere in the page editor, which displays the names of _all_ the objects and sub-objects in use on the page. It would also be where the "Add such-and-such" buttons would go for inline panels.

The bar would probably be on the left, full-height, probably `position:fixed` and just contain a neat list of the items. At the top of the list would be a "Toggle all" button, enabled by default, showing all the objects (in your metaphor, "tracks"). Clicking the nav items below it would either toggle the object's panel into the main view, or switch to it completely i.e only allowing you to see one panel at at time. I hadn't decided which.

The Content, Promote and Settings tabs (or any others added via customisation) would just become headings above the nav items in the list, removing the need for tabs at all.

Ultimately I want to get away from tabs. I particularly want to get away from the limitations to the reasonable number of tabs supported, given their horizontal display. Vertical overflow can just be scrolled. Horizontal overflow is almost always ugly as hell.

If anyone feels like running with this more I'd be really interested to hear any ideas. They might be best discussed on the Wagtail Developers Google Group.

**Comment by danni** [comment]
Can the labels/help_text simply reflow at a certain level of depth in the DOM? I currently have a situation where a ListBlock->StructBlock->StreamBlock is very, very squished.

**Comment by BertrandBordage** [comment]
A temporary solution to drastically improve the layout is simply to partially fix the widths calculus, which is mostly based on percent widths where it should often use `calc()` (I guess the browser support is sufficient now).

You can find a hook fixing the issue here: https://github.com/noripyt/noripyt.com/commit/4e2faa7179dfb8e3c93b83b721582651ea3d499e

## Before
![wagtail-streamfield-before](https://cloud.githubusercontent.com/assets/1119169/16162388/734d69e0-34d3-11e6-9314-4aea82a5c490.png)
## After
![wagtail-streamfield-after](https://cloud.githubusercontent.com/assets/1119169/16162387/734b2996-34d3-11e6-8159-4354ae12216e.png)

**Comment by BertrandBordage** [comment]
Improved workaround for an even better result:
```python
from wagtail.wagtailcore import hooks


# FIXME: Workaround to torchbox/wagtail#1442, remove when fixed.
@hooks.register('insert_editor_css')
def editor_css():
    return """
        <style>
            @media screen and (min-width: 50em) {
                .object fieldset {
                    width: 100%;
                    max-width: 1920px;
                }
                label {
                    max-width: 200px;
                }
            }
            li.sequence-member .struct-block .sequence-container {
                width: 83.3333333%;
            }
            li.sequence-member .sequence-member-inner 
            .sequence-container.sequence-type-list
            > .sequence-container-inner {
                width: 100%;
            }
            li.sequence-member .sequence-member-inner {
                padding: 1.5em;
            }
            .sequence-container > .sequence-container-inner
            > ul.sequence > li.sequence-member > .sequence-member-inner
            > .struct-block > .fields {
                width: 100%;
            }
            .sequence-container.sequence-type-list > .sequence-container-inner
            > ul.sequence > li.sequence-member > .sequence-member-inner
            > .struct-block > .fields {
                width: calc(100% - 100px);
            }
        </style>
    """
```

Screenshot:
![image](https://user-images.githubusercontent.com/1119169/34475458-951bec16-ef8c-11e7-9c5d-0a9a1792c188.png)

**Comment by jmfederico** [comment]
I have a temporary solution for anyone running into this.

Before:
![wagtail-1442-before](https://user-images.githubusercontent.com/237977/63655391-1a599b00-c74d-11e9-8821-81d596738dc8.png)


After:
![wagtail-1442-after](https://user-images.githubusercontent.com/237977/63655393-1d548b80-c74d-11e9-8cd6-385f81d90fd3.png)


```python
from wagtail.core import hooks

@hooks.register('insert_editor_css')
def editor_css_overrides():
    """FIXME: https://github.com/wagtail/wagtail/issues/1442."""
    return """
<style>
/**
 * Start edit overrides
 *
 * Default WagTail styles for Stream and Block editors are not very
 * friendly with nested elements. The width of the elements is reduced
 * for each nesting and end up being unusable.
 *
 * This makes elements use as much width as possible and puts controls
 * on top of the element instead of floating inside them.
 *
 * Also, shows controls only for elements which are hovered.
 * WagTail shows them for hovered elements and its children.
 */

/* Overrides for wagtailadmin/css/core.css */
.field-content {
  width: 100%;
}

/* Overrides for wagtailadmin/css/layouts/page-editor.css */
li.sequence-member .struct-block .fields {
  width: calc(100% - 5px);
}

li.sequence-member .struct-block .sequence-container {
  width: 83.33333%;
}

li.sequence-member
  .sequence-member-inner
  .sequence-type-list
  .sequence-container-inner,
li.sequence-member .struct-block .sequence-container .sequence-container,
li.sequence-member .sequence-member-inner > .sequence .sequence-inner {
  width: 100%;
}

li.sequence-member
  .struct-block
  .sequence-container
  label
  + .sequence-container,
label + div > .field-content {
  width: 83.33333%;
}

/* Move controls for lists */
.sequence-type-list > .sequence-container-inner > .sequence > .sequence-member {
  padding-top: 30px;
}
.list-controls {
  top: 0;
  left: 0;
  right: initial;
}

/* Show controls only for hovered elements, not its children. */
.sequence-member:hover .sequence-controls,
.sequence-member:hover .stream-menu {
  visibility: hidden;
  opacity: 0;
}
.sequence-member:hover > .sequence-controls,
.sequence-member:hover > .stream-menu,
.sequence-container:hover > .stream-menu {
  visibility: visible;
  opacity: 1;
}

/* Make nesting more evident. */
li.sequence-member .sequence-member-inner .sequence-member {
  border: 1px solid transparent;
}
.struct-block li.sequence-member:hover,
li.sequence-member .sequence-member-inner .sequence-member:hover {
  background-color: rgba(217, 227, 227, 0.3);
  border-color: #b0d8d8;
}

/**
 * End edit overrides
 */
</style>
"""
```

**Comment by BertrandBordage** [comment]
We can probably mark this issue as resolved thanks to #5476!
